pallets/tendermint-light-client/src/lib.rs

```diff
@@ -0,0 +1,337 @@
+                        "Cannot finalize light block {:?} because Light Client is not yet initialized",
+                        &untrusted_block,
+                    );
+                    fail!(<Error<T>>::NotInitialized);
+                }
+            };
+
+            let seconds = T::TimeProvider::now().as_secs() as i64;
+            let now = match Time::from_unix_timestamp(seconds, 0) {
+                Ok(now) => now,
+                Err(why) => {
+                    log::error!(
+                        target: "runtime::tendermint-lc",
+                        "Cannot read current time {:?}",
+                        &why,
+                    );
+                    fail!(<Error<T>>::Other)
+                }
+            };
+
+            let to_verify = match untrusted_block.clone().try_into() {
+                Ok(block) => block,
+                Err(why) => {
+                    log::error!(target: "runtime::tendermint-lc", "Conversion failed {:?}", why);
+                    fail!(<Error<T>>::Other);
+                }
+            };
+
+            match verify_light_block::<T>(
+                verifier,
+                &to_verify,
+                &most_recent_trusted_block,
+                &options,
+                now,
+            ) {
+                Verdict::Success => {
+                    match hash {
+                        TendermintHashStorage::Some(hash) => {
+                            // update storage
+                            insert_light_block::<T>(hash, untrusted_block);
+                            log::info!(target: "runtime::tendermint-lc", "Successfully verified light block {:#?}", &hash);
+                            Self::deposit_event(Event::ImportedLightBlock(who, hash));
+                            Ok(())
+                        }
+                        TendermintHashStorage::None => fail!(<Error<T>>::InvalidBlock),
+                    }
+                }
+                Verdict::NotEnoughTrust(voting_power_tally) => {
+                    log::warn!(target: "runtime::tendermint-lc", "Not enough voting power to accept the light block {:#?}, vote tally  {}", &hash, &voting_power_tally);
+                    fail!(<Error<T>>::NotEnoughTrust)
+                }
+                Verdict::Invalid(why) => {
+                    log::warn!(target: "runtime::tendermint-lc", "Rejecting invalid light block {:#?} becasue {}", &hash, &why);
+                    fail!(<Error<T>>::InvalidBlock)
+                }
+            }
+        }
+
+        // TODO: This method will need to be called by the pallet itself if it detects a fork.
+        /// Halt or resume all light client operations
+        ///
+        /// Can only be called by root
+        #[pallet::weight((T::DbWeight::get().reads_writes(1, 1), DispatchClass::Operational))]
+        pub fn set_halted(origin: OriginFor<T>, halted: bool) -> DispatchResult {
+            ensure_root(origin)?;
+            <IsHalted<T>>::put(halted);
```

Comment:
Let's check first if the current `IsHalted` isn't already equal to `halted`. In such case, this should be a no-op rather than logging and creating new events.

Reply:
I introduced a pattern match and log whether there was or was not any operation. TODO : weights need ot be adjusted but that's generally true for all calls, so I will include it in the benchmarks PR

Reply:
It's not only about the cost but I think it might be confusing to people when they see an event about halting the light client. 

Not a blocker of course, we can leave it as is if you prefer.